fido/Crypto/Fido2/Attestation/AndroidKey.hs

```diff
@@ -0,0 +1,130 @@
+import Data.Binary (Binary, decode)
+import Data.ByteArray (convert)
+import Data.ByteString.Lazy (fromStrict)
+import Data.X509 (Certificate (certPubKey), getCertificate)
+import qualified Data.X509 as X509
+
+-- https://android.googlesource.com/platform/hardware/libhardware/+/master/include/hardware/keymaster_defs.h
+kmOriginGenerated :: Integer
+kmOriginGenerated = 0
+
+-- https://android.googlesource.com/platform/hardware/libhardware/+/master/include/hardware/keymaster_defs.h
+kmPurposeSign :: Integer
+kmPurposeSign = 2
+
+-- https://www.w3.org/TR/webauthn-2/#sctn-packed-attestation
+verify :: Stmt -> AuthenticatorData -> Digest SHA256 -> Either AttestationError AttestedCredentialData
+verify Stmt {alg = _alg, sig, x5c} AuthenticatorData {rawData, attestedCredentialData} clientDataHash = do
+  -- 1. Verify that attStmt is valid CBOR conforming to the syntax defined above and perform CBOR decoding on it to
+  -- extract the contained fields.
+  -- NOTE: The validity of the data is already checked during decoding.
+
+  -- 2. Verify that sig is a valid signature over the concatenation of authenticatorData and clientDataHash using the
+  -- public key in the first certificate in x5c with the algorithm specified in alg.
+  -- TODO: Maybe use verifyX509Sig like in Packed.hs
+  let signedData = rawData <> convert clientDataHash
+      cert = getCertificate x5c
+  x5cKey <- maybe (Left CertiticatePublicKeyInvalid) pure $ PublicKey.toPublicKey $ certPubKey cert
+  unless (PublicKey.verify x5cKey signedData sig) $ Left StatementInvalidSignature
+
+  -- 3. Verify that the public key in the first certificate in x5c matches the credentialPublicKey in the
+  -- attestedCredentialData in authenticatorData.
+  credData <- maybe (Left AttestationCredentialDataMissing) pure attestedCredentialData
+  let key = credentialPublicKey credData
+  unless (key == x5cKey) $ Left AttestationCredentialKeyMismatch
+
+  -- 4. Verify that the attestationChallenge field in the attestation certificate extension data is identical to
+  -- clientDataHash.
+  -- See https://source.android.com/security/keystore/attestation for the ASN1 description
+  let (X509.Extensions mX509Exts) = X509.certExtensions cert
+      mX509Ext = mX509Exts >>= findProperExtension [1, 3, 6, 1, 4, 1, 11129, 2, 1, 17]
+  ext <- maybe (Left AttestationRecordExtensionMissing) pure mX509Ext
+  asn1 <- either (Left . ASN1Error) pure . decodeASN1 DER $ fromStrict (X509.extRawContent ext)
+  attestationChallenge <- decodeAttestationChallenge asn1
+  unless (attestationChallenge == clientDataHash) $ Left AttestationChallengeMismatch
+
+  -- 5. Verify the following using the appropriate authorization list from the attestation certificate extension data:
+
+  -- 5.a The AuthorizationList.allApplications field is not present on either
+  -- authorization list (softwareEnforced nor teeEnforced), since
+  -- PublicKeyCredential MUST be scoped to the RP ID.
+  let asnCollections = drop 6 asn1
```

Comment:
Seemingly arbitrary constants aren't very nice. Same for the 702, 1 and 600 later down. Instead of this we should write a separate parser for the ASN.1 structure, potentially only storing the fields we're interested in, and then access those fields here.